lib/membrane_audio_interleaver.ex

```diff
@@ -0,0 +1,283 @@
+      {new_queue_size, state} = add_payload(payload, pad, state)
+
+      if new_queue_size >= Caps.sample_size(caps) do
+        interleave_and_return(state)
+      else
+        {{:ok, redemand: :output}, state}
+      end
+    end
+  end
+
+  @impl true
+  def handle_caps(pad, caps, _context, state) do
+    case state.caps do
+      nil ->
+        state = %{state | caps: caps}
+        {{:ok, caps: {:output, %{caps | channels: 2}}, redemand: :output}, state}
+
+      ^caps ->
+        {:ok, state}
+
+      _invalid_caps ->
+        raise(
+          RuntimeError,
+          "received invalid caps on pad #{inspect(pad)}, expected: #{inspect(state.caps)}, got: #{inspect(caps)}"
+        )
+    end
+  end
+
+  # send demand to input pads where current queue is not long enough
+  defp do_handle_demand(size, %{pads: pads} = state) do
+    if state.finished do
+      {:ok, state}
+    else
+      demands =
+        Enum.map(
+          pads,
+          fn {pad, %{queue: queue}} ->
+            demand_size =
+              queue
+              |> byte_size()
+              |> then(&max(0, size - &1))
+
+            {:demand, {pad, demand_size}}
+          end
+        )
+
+      {{:ok, demands}, state}
+    end
+  end
+
+  # try to interleave channels and formulate proper element callback return message
+  defp interleave_and_return(state) do
+    case(try_interleave(state)) do
+      :none -> {:ok, state}
+      {:finished, {buffer, state}} -> {{:ok, buffer: buffer, end_of_stream: :output}, state}
+      {:ok, {buffer, state}} -> {{:ok, buffer: buffer}, state}
+    end
+  end
+
+  # interleave channels only if all queues are long enough (have at least 'sample_size' size)
+  defp try_interleave(%{caps: caps, pads: pads, order: order} = state) do
+    sample_size = Caps.sample_size(caps)
+
+    min_length =
+      min_queue_length(pads)
+      |> trunc_to_whole_samples(sample_size)
+
+    if min_length >= sample_size do
+      {payload, pads} = DoInterleave.interleave(min_length, sample_size, pads, order)
+      state = %{state | pads: pads}
+      buffer = {:output, %Buffer{payload: payload}}
+
+      if any_finished?(pads, sample_size) do
+        {:finished, {buffer, %{state | finished: true}}}
+      else
+        {:ok, {buffer, state}}
+      end
+    else
+      :none
+    end
+  end
+
+  # Returns minimum number of bytes present in all queues
+  defp min_queue_length(pads) do
+    pads
+    |> Enum.map(fn {_pad, %{queue: queue}} -> byte_size(queue) end)
+    |> Enum.min(fn -> 0 end)
+  end
+
+  # Returns the biggest multiple of `sample_size` that is not bigger than `size`
+  defp trunc_to_whole_samples(size, sample_size)
+       when is_integer(size) and is_integer(sample_size) do
+    rest = rem(size, sample_size)
+    size - rest
+  end
+
+  defp any_finished?(pads, sample_size) do
+    pads
+    |> Enum.any?(fn
+      {_pad, %{queue: queue, stream_ended: true}} when byte_size(queue) < sample_size -> true
+      _entry -> false
+    end)
```

Comment:
- Using pipe operator once should be avoided.
- Ended stream and `byte_size(queue) < sample_size` means that the input is invalid, so maybe it would be better to raise error in such situation